azure/services/aso/tags.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aso
+
+import (
+	"encoding/json"
+
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/services/tags"
+	"sigs.k8s.io/cluster-api-provider-azure/util/maps"
+)
+
+// tagsLastAppliedAnnotation is the key for the annotation which tracks the AdditionalTags.
+// See https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
+// for annotation formatting rules.
+const tagsLastAppliedAnnotation = "sigs.k8s.io/cluster-api-provider-azure-last-applied-tags"
+
+// reconcileTags modifies parameters in place to update its tags and its last-applied annotation.
+func reconcileTags(t TagsGetterSetter, existing genruntime.MetaObject, parameters genruntime.MetaObject) error {
+	lastAppliedTags := map[string]interface{}{}
+	if existing != nil {
+		lastAppliedTagsJSON := existing.GetAnnotations()[tagsLastAppliedAnnotation]
+		if lastAppliedTagsJSON != "" {
+			err := json.Unmarshal([]byte(lastAppliedTagsJSON), &lastAppliedTags)
+			if err != nil {
+				return errors.Wrapf(err, "failed to unmarshal JSON from %s annotation", tagsLastAppliedAnnotation)
+			}
+		}
+	}
+
+	existingTags := t.GetActualTags(existing)
+	existingTagsMap := converters.TagsToMap(existingTags)
+
+	_, createdOrUpdated, deleted, newAnnotation := tags.TagsChanged(lastAppliedTags, t.GetAdditionalTags(), existingTagsMap)
+	newTags := maps.Merge(maps.Merge(existingTags, t.GetDesiredTags(parameters)), createdOrUpdated)
```

Comment:
Is there a difference between the use of `t.GetDesiredTags(parameters)` and `t.GetAdditionalTags()`? It seems like both are referring to the desired tags in this case, but one refers to the AdditionalTags field and the other is the tags declared in the spec.

Reply:
I'm just a bit confused since we are comparing lastAppliedTags to additionalTags at the line above, but then merging the existingtags with the tags from parameters.

Reply:
Ah that makes sense. In this case, GetDesiredTags gets tags that a user has specified by editing the ASO resource directly, and not through the portal or CLI correct?

Reply:
Correct